### DISCUSSION TITLE
Type of Tesla Application Access

### DISCUSSION THREAD

**Discussion by IanK6449** [llamafilm/tesla-http-proxy-addon#61] [General]
Hi,

I am trying to get the Tesla HTTP Proxy to work and have got to the point where I can view the public key provided by NGINX.
The next step is to apply for the Tesla API access but the documentation doesn't specify whether I need to request access for machine and user or just machine to machine? I ask as, if just requesting machine to machine, I am not asked to enter any URLs / URI's whereas if requesting for user and machine I am.
However, when I apply for just machine to machine, I am given an ID and secret which I enter into the Tesla HTTP Proxy configuration and try to start the proxy. At this point I get the error message below. However, if I try to request user and machine access instead I am told that my URL is not certified even though I can see the public key at the "tesla.example.com"/.well-known/appspecific/com.tesla.3p.public-key.pem address given in the documentation.
So, my question is, which type of Tesla API access do I need to request and, if the user and machine type, does anybody have a clue why I keep getting the URL does not have a valid cert when I can access it easily as per the instructions?
I suspect there is something wrong with my NGINX Proxy setup perhaps but I don't know how to check whether it includes both my HA domain name AND my "example.tesla.com" domain name or just one of them which may explain the URL cert errors I am getting from the Tesla API access request.

Error message when using the machine to machine API application:

[11:24:28] INFO: Running auth.py
[11:24:29] auth:INFO: Generating Partner Authentication Token
[11:24:29] urllib3.connectionpool:DEBUG: Starting new HTTPS connection (1): auth.tesla.com:443
[11:24:30] urllib3.connectionpool:DEBUG: https://auth.tesla.com:443 "POST /oauth2/v3/token HTTP/1.1" 200 763
[11:24:30] auth:DEBUG: {"access_token":"_REDACTED_","expires_in":28800,"token_type":"Bearer"}
[11:24:30] auth:INFO: Registering Tesla account...
[11:24:30] urllib3.connectionpool:DEBUG: Starting new HTTPS connection (1): fleet-api.prd.na.vn.cloud.tesla.com:443
[11:24:30] urllib3.connectionpool:DEBUG: https://fleet-api.prd.na.vn.cloud.tesla.com:443 "POST /api/1/partner_accounts HTTP/1.1" 412 217
[11:24:30] auth:ERROR: Error 412: {"response":null,"error":"Root domain _"Tesla.Example.Com"_ must match registered allowed origin on https://developer.tesla.com/dashboard/app-details/6610f3b44de4-45b1-9d1a-fbcb5485e2c2","error_description":""}
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service webui: stopping
s6-rc: info: service webui successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped

**Comment by llamafilm** [comment]
Good point, I should update the docs to say you need user and machine access.

Lots of people have trouble getting their domain validated. There are essentially 2 reasons, either your public key is not accessible by their server, or your TLS certificate is not trusted by their server.  If you share your domain here maybe someone can help.

This key is, by definition, public, so there’s no reason to hide it.

**Reply by IanK6449** [reply]
Thanks for a very speedy reply!
Yeah the public key is, by definition, public so here is the path:
https://weerasinghetesla.duckdns.org/.well-known/appspecific/com.tesla.3p.public-key.pem
I am an absolute dummy when it comes to TLS and certs so any assistance you can give me would be hugely appreciated. It's taken me several hours to even get to the point where I can see the cert.

One thing that may be helpful is that Chrome reports that link as a dangerous site yet the cert is valid. So, maybe the cert is missing some detail?

Oh and yeah, a quick documentation update may not go awry :-) But I think we all appreciate that this isn't your day job so thank you for finding the time to build this component!

**Reply by llamafilm** [reply]
I can't reach that URL at all.  It resolves to 167.179.178.177 but doesn't seem to be alive.  Have you tested it from outside your house?  Perhaps your firewall is blocking something.

**Reply by IanK6449** [reply]
Ah, so trying it from outside my network it is, indeed, unreachable. I currently have port 443 forwarded to 443 but I am wondering whether this is correct? I am also using the AdGuard addon so I'll try disabling that and seeing if it renders the public key accessible from outside.

**Reply by IanK6449** [reply]
Also, I had a DuckDNS name setup to reach that IP address but it is the same IP address as my other DuckDNS entry to normally reach my HASS instance remotely. Have I set this up correctly, i.e. should I have 2 DuckDNS entries or just the one? If just the one, I can't see how my Tesla domain name would get resolved by any external domain (i.e. the Tesla API application)? Or does NGINX add this DNS entry somewhere in the cloud?
As you can no doubt tell, this security stuff is a dark art to me.
